### PR TITLE
UNOMI-817: set pastEvents as nested and clean existing mapping

### PR DESCRIPTION
**Pull request by jsinovassin**
https://issues.apache.org/jira/browse/UNOMI-817

Set the pastEvents properties as nested in the mapping of the profile.
Creation of a migration script to clean the mapping of the existing profiles.
Do not copy anymore the pastEvents in the session when creating the session.

### CODE REVIEW

**Review by jkevan — CHANGES_REQUESTED** [review]

File: tools/shell-commands/src/main/resources/requestBody/2.5.0/update_pastEvents_profile.painless [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+if (ctx._source.systemProperties != null && ctx._source.systemProperties.pastEvents != null && ctx._source.systemProperties.pastEvents instanceof Map) {
+    Map updatedPastEvents = new HashMap();
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply tools/shell-commands/src/main/resources/requestBody/2.5.0/update_pastEvents_profile.painless:19]: Done!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java [hunk 2]
```diff
@@ -123,14 +121,27 @@ public int execute(Action action, Event event) {
...
+        ArrayList<Map<String, Object>> existingPastEvents = (ArrayList<Map<String, Object>>) event.getProfile().getSystemProperties().get("pastEvents");
+        if (existingPastEvents != null) {
+            pastEvents.addAll(existingPastEvents.stream().filter(pastEvent -> !pastEvent.get("key").equals(generatedPropertyKey)).collect(Collectors.toList()));
+        }
+
+        Map<String, Object> pastEvent = new HashMap<>();
+        pastEvent.put("key", generatedPropertyKey);
+        pastEvent.put("count", count);
+        pastEvents.add(pastEvent);
+        return PropertyHelper.setProperty(event.getProfile(), "systemProperties.pastEvents", pastEvents, "alwaysSet");
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java:142]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java [hunk 3]
```diff
@@ -123,14 +121,27 @@ public int execute(Action action, Event event) {
...
+        if (updatePastEvents(event, (String) pastEventCondition.getParameter("generatedPropertyKey"), count)) {
             return EventService.PROFILE_UPDATED;
         }
 
         return EventService.NO_CHANGE;
     }
 
+    private boolean updatePastEvents(Event event, String generatedPropertyKey, long count) {
+        ArrayList<Map<String, Object>> pastEvents = new ArrayList<>();
+        ArrayList<Map<String, Object>> existingPastEvents = (ArrayList<Map<String, Object>>) event.getProfile().getSystemProperties().get("pastEvents");
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java:133]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java [hunk 4]
```diff
@@ -131,35 +131,77 @@ private Condition getProfileIdsCondition(Set<String> ids, boolean shouldMatch) {
...
+
+            // 1. Key not present in profile
+            Condition keyNestedCondition = new Condition();
+            keyNestedCondition.setConditionType(definitionsService.getConditionType("nestedCondition"));
+            keyNestedCondition.setParameter("path", "systemProperties.pastEvents");
+
+            Condition keyEquals = new Condition(profilePropertyConditionType);
+            keyEquals.setParameter("propertyName", "systemProperties.pastEvents.key");
+            keyEquals.setParameter("comparisonOperator", "equals");
+            keyEquals.setParameter("propertyValue", generatedPropertyKey);
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java:166]: keyEquals was added as a subCondition of keyNestedCondition at line 168

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java [hunk 5]
```diff
@@ -131,35 +131,77 @@ private Condition getProfileIdsCondition(Set<String> ids, boolean shouldMatch) {
     }
 
     private Condition getProfileConditionForCounter(String generatedPropertyKey, Integer minimumEventCount, Integer maximumEventCount, boolean eventsOccurred) {
-        String generatedPropertyName = "systemProperties.pastEvents." + generatedPropertyKey;
+        Condition countCondition = new Condition();
+
+        countCondition.setConditionType(definitionsService.getConditionType("nestedCondition"));
+        countCondition.setParameter("path", "systemProperties.pastEvents");
+
+        Condition subConditionCount = new Condition(definitionsService.getConditionType("profilePropertyCondition"));
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java:139]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java [hunk 6]
```diff
@@ -131,35 +131,77 @@ private Condition getProfileIdsCondition(Set<String> ids, boolean shouldMatch) {
...
     }
 
+    private Condition getKeyEqualsCondition(String generatedPropertyKey) {
+        Condition subConditionKey = new Condition(definitionsService.getConditionType("profilePropertyCondition"));
+        subConditionKey.setParameter("propertyName", "systemProperties.pastEvents.key");
+        subConditionKey.setParameter("comparisonOperator", "equals");
+        subConditionKey.setParameter("propertyValue", generatedPropertyKey);
+        return subConditionKey;
+    }
+
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java:204]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionEvaluator.java [hunk 7]
```diff
@@ -55,10 +57,15 @@ public boolean eval(Condition condition, Item item, Map<String, Object> context,
         if (parameters.containsKey("generatedPropertyKey")) {
             String key = (String) parameters.get("generatedPropertyKey");
             Profile profile = (Profile) item;
-            Map<String,Object> pastEvents = (Map<String, Object>) profile.getSystemProperties().get("pastEvents");
-            if (pastEvents != null) {
-                Number l = (Number) pastEvents.get(key);
-                count = l != null ? l.longValue() : 0L;
+            Object rawPastEvents =  profile.getSystemProperties().get("pastEvents");
+            if (rawPastEvents != null) {
+                List<Map<String, Object>> pastEvents = (ArrayList<Map<String, Object>>) rawPastEvents;
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionEvaluator.java:62]: Updated!

**Review by jkevan — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/apache/unomi/api/utils/ConditionHelper.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+import org.apache.unomi.api.services.DefinitionsService;
+
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Utility class for creating various types of {@link Condition} objects.
+ * This class provides methods to easily construct conditions used for querying data based on specific criteria.
+ */
+public class ConditionHelper {
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply api/src/main/java/org/apache/unomi/api/utils/ConditionHelper.java]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java [hunk 9]
```diff
@@ -123,14 +121,38 @@ public int execute(Action action, Event event) {
...
+        return addNewPastEvent(existingPastEvents, generatedPropertyKey, count);
+    }
+
+    private boolean addNewPastEvent(List<Map<String, Object>> existingPastEvents, String key, long count) {
+        Map<String, Object> newPastEvent = new HashMap<>();
+        newPastEvent.put("key", key);
+        newPastEvent.put("count", count);
+        existingPastEvents.add(newPastEvent);
+        return true;  // New event added
+    }
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java [hunk 10]
```diff
@@ -122,44 +128,42 @@ protected static boolean getStrategyFromOperator(String operator) {
...
+    private Condition createEventOccurredCondition(String generatedPropertyKey, Integer minimumEventCount, Integer maximumEventCount) {
+        Condition subConditionCount = conditionHelper.createProfilePropertyCondition("systemProperties.pastEvents.count", "between", Arrays.asList(minimumEventCount, maximumEventCount), "propertyValuesInteger");
+        Condition subConditionKey = conditionHelper.createProfilePropertyCondition("systemProperties.pastEvents.key", "equals", generatedPropertyKey, "propertyValue");
+
+        Condition booleanCondition = conditionHelper.createBooleanCondition("and", Arrays.asList(subConditionCount, subConditionKey));
+        return conditionHelper.createNestedCondition("systemProperties.pastEvents", booleanCondition);
+    }
+
+    private Condition createEventNotOccurredCondition(String generatedPropertyKey) {
+        Condition pastEventsNotExist = createPastEventMustNotExistCondition(generatedPropertyKey);
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java]: Updated!

File: services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java [hunk 11]
```diff
@@ -986,31 +986,35 @@ public void recalculatePastEventConditions() {
...
+
+
+            if (batch.size() == segmentUpdateBatchSize || (!entryIterator.hasNext() && !batch.isEmpty())) {
                 try {
-                    persistenceService.update(batch, Profile.class);
+                    batch.forEach((id, params) -> {
+                        Condition profileIdCondition = conditionHelper.createProfilePropertyCondition("itemId", "equals", id, "propertyValue");
+                        Condition[] conditions = new Condition[]{profileIdCondition};
+                        persistenceService.updateWithQueryAndStoredScript(Profile.class, new String[]{"updatePastEventOccurences"}, params, conditions);
+                    });
```
[the remark this answers is not in the document]
Comment (jsinovassin) [reply services/src/main/java/org/apache/unomi/services/impl/segments/SegmentServiceImpl.java]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java [hunk 12]
```diff
@@ -123,14 +121,37 @@ public int execute(Action action, Event event) {
...
+    private boolean updatePastEvents(Event event, String generatedPropertyKey, long count) {
+        List<Map<String, Object>> existingPastEvents = (List<Map<String, Object>>) event.getProfile().getSystemProperties().get("pastEvents");
+        if (existingPastEvents == null) {
+            existingPastEvents = new ArrayList<>();
+            event.getProfile().getSystemProperties().put("pastEvents", existingPastEvents);
+        }
+
+        for (Map<String, Object> pastEvent : existingPastEvents) {
+            if (generatedPropertyKey.equals(pastEvent.get("key"))) {
+                if (pastEvent.get("count").equals(count)) {
```
Comment (jkevan) [inline plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java:140]: `pastEvent.get("count")` does it need NPE check ?

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java [hunk 13]
```diff
@@ -122,44 +124,39 @@ protected static boolean getStrategyFromOperator(String operator) {
...
+        ConditionBuilder.ConditionItem keyExistsAndCounterZero = conditionBuilder.and(counterZero, keyEquals);
+        ConditionBuilder.ConditionItem nestedKeyExistsAndCounterZero = conditionBuilder.nested(keyExistsAndCounterZero, "systemProperties.pastEvents");
+        return conditionBuilder.or(counterMissing, nestedKeyExistsAndCounterZero).build();
+    }
+
+    private ConditionBuilder.ConditionItem createPastEventMustNotExistCondition(String generatedPropertyKey) {
+        ConditionBuilder conditionBuilder = definitionsService.getConditionBuilder();
+        ConditionBuilder.ConditionItem keyEquals = conditionBuilder.profileProperty("systemProperties.pastEvents.key").equalTo(generatedPropertyKey);
+        ConditionBuilder.ConditionItem keyNestedCondition = conditionBuilder.nested(keyEquals, "systemProperties.pastEvents");
+        return conditionBuilder.not(keyNestedCondition);
```
Comment (jkevan) [inline plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java:157]: To be verified, but I'm not sure we required a nested level here.
Nested query is useful to do inner conditioning on a sub level of a nested object using multiple clauses, those multiple clauses will then be evaluated on per nested item basis. 
Like for condition on both: `systemProperties.pastEvents.count` and `systemProperties.pastEvents.key`

But here we only have one clause one `systemProperties.pastEvents.key`, so I wonder if nested level is really useful or not. 
To my current understanding of nested ES queries (a bit rusty) it's seem's to me we could do a simple:

`conditionBuilder.not(conditionBuilder.profileProperty("systemProperties.pastEvents.key").equalTo(generatedPropertyKey))`

**Review by jkevan — APPROVED** [review]

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java (hunk 12 above)
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/actions/SetEventOccurenceCountAction.java]: Updated!

File: plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java (hunk 13 above)
Comment (jsinovassin) [reply plugins/baseplugin/src/main/java/org/apache/unomi/plugins/baseplugin/conditions/PastEventConditionESQueryBuilder.java]: Updated!

**Pull request merged**